vulkan/vulkan_application.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/vulkan/vulkan_application.h"
+#include <utility>
+#include <vector>
+#include "flutter/vulkan/vulkan_device.h"
+#include "flutter/vulkan/vulkan_proc_table.h"
+#include "flutter/vulkan/vulkan_utilities.h"
+
+namespace vulkan {
+
+VulkanApplication::VulkanApplication(
+    VulkanProcTable& p_vk,
+    const std::string& application_name,
+    std::vector<std::string> enabled_extensions,
+    uint32_t application_version,
+    uint32_t api_version)
+    : vk(p_vk), api_version_(api_version), valid_(false) {
+  // Check if we want to enable debugging.
+
+  bool enable_instance_debugging =
+      EnableDebugging() && VulkanDebugReport::DebugExtensionSupported(vk);
+
+  // Configure extensions.
+
+  if (enable_instance_debugging) {
+    enabled_extensions.emplace_back(VulkanDebugReport::DebugExtensionName());
+  }
+
+  const char* extensions[enabled_extensions.size()];
+
+  for (size_t i = 0; i < enabled_extensions.size(); i++) {
+    extensions[i] = enabled_extensions[i].c_str();
+  }
+
+  // Configure layers.
+
+  const std::vector<std::string> enabled_layers = InstanceLayersToEnable(vk);
+
+  const char* layers[enabled_layers.size()];
+
+  for (size_t i = 0; i < enabled_layers.size(); i++) {
+    layers[i] = enabled_layers[i].c_str();
+  }
+
+  // Configure init structs.
+
+  const VkApplicationInfo info = {
+      .sType = VK_STRUCTURE_TYPE_APPLICATION_INFO,
+      .pNext = nullptr,
+      .pApplicationName = application_name.c_str(),
+      .applicationVersion = application_version,
+      .pEngineName = "FlutterEngine",
+      .engineVersion = VK_MAKE_VERSION(1, 0, 0),
+      .apiVersion = api_version_,
+  };
+
+  const VkInstanceCreateInfo create_info = {
+      .sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
+      .pNext = nullptr,
+      .flags = 0,
+      .pApplicationInfo = &info,
+      .enabledLayerCount = static_cast<uint32_t>(enabled_layers.size()),
+      .ppEnabledLayerNames = layers,
+      .enabledExtensionCount = static_cast<uint32_t>(enabled_extensions.size()),
+      .ppEnabledExtensionNames = extensions,
+  };
+
+  // Perform initialization.
+
+  VkInstance instance = VK_NULL_HANDLE;
+
+  if (VK_CALL_LOG_ERROR(vk.CreateInstance(&create_info, nullptr, &instance)) !=
+      VK_SUCCESS) {
+    return;
+  }
+
+  // Now that we have an instance, setup instance proc table entries.
+  if (!vk.SetupInstanceProcAddresses(instance)) {
+    return;
+  }
+
+  instance_ = {instance, [this](VkInstance i) {
+                 FTL_LOG(INFO) << "Destroying instance";
```

Comment:
> Also, it seems a little dodgy to capture "this" in a closure that will be triggered to run by the destruction of "this". In particular, it relies on the declaration order of instance fields in the header. Could you capture "p_vk" instead?

I really struggled with this part of the API. I need access to proc table everywhere but did not want to make it a `RefPtr` to avoid the overhead of the reference counting. So I just settled on the convention of having the proc table be the first instance field of any class that requires it past the constructor with the caller providing the guarantee that it is responsible for the lifecycle of the both the proc table and the instance being created (and won't let the instance be collected before the proc table). Given this guarantee, is this behavior clearer? I suppose I could also make the proc table reference as an instance field be create using a macro or something that asserts that it is the first one in the class.